train/unittest_train.py

```diff
@@ -0,0 +1,184 @@
+        save_model(self.model, self.unittest_data_path, self.output_model_file)
+
+    @classmethod
+    def tearDownClass(self):
+        ''' Clean up function '''
+        shutil.rmtree(self.unittest_data_path)
+
+    def test_df_read(self):
+        ''' Checks if pandas dataframe is retuned '''
+        self.assertIsInstance(
+            self.df, pandas.core.frame.DataFrame
+        )
+
+    def test_df_attributes(self):
+        ''' Checks the pandas dataframe attributes '''
+        self.assertListEqual(
+            list(self.df.columns.values), ["target", "timestamp", "datetime", "query", "user", "text"]
+        )
+
+    def test_preprocess_text(self):
+        ''' Checks pre-processing of texts '''
+        self.assertEqual(
+            self.test_cfg['processed_text'], preprocess(self.test_cfg['raw_text'], self.stemmer)
+        )
+      
+    def test_sentiment_mapping(self):
+        ''' Checks the sentiment mapping function '''
+        self.assertEqual(
+            "NEGATIVE", decode_sentiment(0)
+        )
+        self.assertEqual(
+            "NEUTRAL", decode_sentiment(2)
+        )
+        self.assertEqual(
+            "POSITIVE", decode_sentiment(4)
+        )
+
+    def test_split_train_test_data(self):
+        ''' Checks the train and test dataset ratio '''
+        self.assertAlmostEqual(
+            len(self.df_train) / len(self.df), self.train_size_ratio
+        )
+        self.assertAlmostEqual(
+            len(self.df_test) / len(self.df), 1 - self.train_size_ratio
+        )
+
+    def test_docs_creation(self):
+        ''' Checks the document creation process '''
+        self.assertListEqual(
+            create_documents(self.create_docs_df)[0], self.test_cfg['document'] 
+        )
+
+    def test_labels_generated(self):
+        ''' Checks the labels generated '''
+        self.assertListEqual(
+            sorted(["POSITIVE", "NEGATIVE", "NEUTRAL"]), sorted(self.labels)
+        )
+        
+    def test_model_predictions(self):
+        ''' Checks the model performance metrics - loss and accuracy '''
+        # Checks the model accuracy
+        self.assertTrue(
+            0 < self.score[1] <= 1
+        )
+        # Checks the model loss
+        self.assertTrue(
+            0 < self.score[0] 
+        )
+
+    def test_model_history(self):
+        ''' Checks if loss and accuracy are returned for all of the epochs specified above '''
+        self.assertEqual(
+            len(self.acc), self.epochs_val
+        )
+        self.assertEqual(
+            len(self.val_acc), self.epochs_val
+        )
+        self.assertEqual(
+            len(self.loss), self.epochs_val
+        )
+        self.assertEqual(
+            len(self.val_loss), self.epochs_val
+        )
+
+    def test_model_history_metrics(self):
+        ''' Checks the average performance metrics from history '''
+        self.assertTrue(
+            0 < sum(self.acc) / len(self.acc) <= 1
```

Comment:
same here the lower bound is (0) which too loose. 

Also is this an average? or something else?

Reply:
@orionsBeltWest Yes, I guess last value is good (already done)